xdsl/transforms/eqsat_serialize_egraph.py

```diff
@@ -0,0 +1,69 @@
+import json
+
+from xdsl.context import Context
+from xdsl.dialects import builtin, eqsat, func
+from xdsl.ir import BlockArgument, Operation
+from xdsl.passes import ModulePass
+
+
+def serialize_func_to_egraph(f_op: func.FuncOp):
+    assert len(f_op.regions) == 1
+    body = f_op.regions[0]
+    assert body.first_block == body.last_block, (
+        "Only single block functions are supported"
+    )
+    body = body.first_block
+
+    i_eclass = 0
+    i_enode = 0
+    eclass_to_id: dict[eqsat.EClassOp, int] = {}
+    op_to_id = {}
+    nodes: list[dict[str, str | list[str]]] = []
+
+    if body:
+        for op in body.ops:
+            if isinstance(op, eqsat.EClassOp):
+                eclass_to_id[op] = (
+                    i_enode  # pick the first e-node as the representative
+                )
+                assert len(op.operands) > 0, "EClassOp must have operands"
+                for operand in op.operands:
+                    node = operand.owner
+                    children: list[str] = []
+                    if isinstance(node, Operation):
+                        op_to_id[node] = i_enode
+                        if len(node.operands) > 0:
+                            name = node.name
+                        else:
+                            name = str(node).split("=")[1].split(":")[0].strip()
+                        for child_op in node.operands:
+                            assert isinstance(child_op.owner, eqsat.EClassOp)
+                            child = eclass_to_id.get(child_op.owner)
+                            assert child is not None, "Child eclass not found"
+                            children.append(f"enode_{child}")
+                    else:
+                        assert isinstance(operand, BlockArgument)
+                        op_to_id[operand] = i_enode
+                        name = f"arg {operand.index}"
+
+                    nodes.append(
+                        {
+                            "op": name,
+                            "eclass": f"eclass_{i_eclass}",
+                            "children": children,
+                        }
+                    )
+                    i_enode += 1
+                i_eclass += 1
+        print(
+            json.dumps({"nodes": {f"enode_{i}": node for i, node in enumerate(nodes)}})
+        )
```

Comment:
Let's decouple the dictionary construction from json printing? It would be good to have two functions